src/fullfile.c

```diff
@@ -139,15 +139,30 @@ int download_fullfiles(struct list *files, int *num_downloads)
 	/* make a new list with only the files we actually need to download */
 	for (iter = list_head(files); iter; iter = iter->next) {
 		char *targetfile;
+		char *targetfile_cache;
 		file = iter->data;
 
 		if (file->is_deleted || file->do_not_update) {
 			continue;
 		}
 
+		/* Check the statedir and statedir-cache for the expected fullfile. When the
+		 * fullfile exists in the statedir-cache, it is copied to the statedir. */
 		string_or_die(&targetfile, "%s/staged/%s", globals.state_dir, file->hash);
 		if (lstat(targetfile, &stat) != 0 || !verify_file(file, targetfile)) {
-			need_download = list_append_data(need_download, file);
+			bool download = true;
```

Comment:
this boolean here makes the code a bit hard to read. Can you group all this block (including the if lstat) in a function and return if we need to download or not?

    if (!fullfile_exists_or_cached()) {
        need_download = list_append_data(need_download, file);
    }